File: main/redundant-connection/redundant-connection.cpp
namespace {
    class DisjointSets {
    public:
        explicit DisjointSets(const size_t size) : elems_(size, -1) { }

        // If i and j are in different sets, joins the sets and returns true.
        // If they are in the same set already, just returns false.
        bool unite(int i, int j) noexcept;

    private:
        int find(int i) noexcept;

        vector<int> elems_;
    };

    bool DisjointSets::unite(int i, int j) noexcept
    {
        // Find the ancestors and check if they're already the same.
        i = find(i);
        j = find(j);
        if (i == j) return false;

        // Union by rank.
        if (elems_[i] > elems_[j]) {
            // j has superior (more negative) rank.
            elems_[i] = j;
        } else {
            // If i and j have the same rank, promote i.
            if (elems_[i] == elems_[j]) --elems_[i];

            // i has superior (more negative) rank.
            elems_[j] = i;
        }

        return true;
    }

    int DisjointSets::find(int i) noexcept
    {
        // Find the ancestor.
        auto j = i;
        while (elems_[j] >= 0) j = elems_[j];

        // Compress the path.
        while (i != j) {
            const auto p = elems_[i];
            elems_[i] = j;
            i = p;
        }

        return j;
    }
}

class Solution {
public:
    static vector<int>
    findRedundantConnection(const vector<vector<int>>& edges);
};

vector<int> Solution::findRedundantConnection(const vector<vector<int>>& edges)
{
    DisjointSets sets {edges.size() + 1u}; // because we're not using vertex 0

    for (const auto& edge : edges)
        if (!sets.unite(edge.front(), edge.back())) return edge;

    return {};
}

File: main/redundant-connection/redundant_connection.rb
# Disjiont set data structure with union by rank and path compression.
class DisjointSets
  def initialize(size)
    @elems = Array.new(size, -1)
  end

  def union(i, j)
    # Find the ancestors and check if they are already in the same set.
    i = find(i)
    j = find(j)
    return false if i == j

    # Unite the sets by rank.
    if @elems[i] > @elems[j]
      # j has superior (more negative) rank.
      @elems[i] = j
    else
      # If i and j have the same rank, promote i.
      @elems[i] -= 1 if @elems[i] == @elems[j]

      # i has superior (more negative) rank.
      @elems[j] = i
    end

    true
  end

  private def find(i)
    # Find the ancestor.
    j = i
    j = @elems[j] while @elems[j] >= 0

    # Compress the path.
    while i != j
      p = @elems[i]
      @elems[i] = j
      i = p
    end

    j
  end
end

# @param {Integer[][]} edges
# @return {Integer[]}
def find_redundant_connection(edges)
  sets = DisjointSets.new(edges.length + 1) # because we're not using 0

  edges.each do |edge|
    return edge unless sets.union(*edge)
  end

  nil
end
